### DISCUSSION TITLE
Gitignore not working with terraform

### DISCUSSION THREAD

**Discussion by mkx85** [community/community#23052] [New to GitHub]
<p>I am new to github and I’ve tried a few things to successfully push my code to my github account but it won’t work.</p>
<p>I’m using Visual studio code and want to push some basic Terraform files to my repo. When I go to push the project, it complains about the large file in the .terraform file. But the terraform directory is clearly excluded in the .gitignore file. I’ve done the following:</p>
<ol>
<li>Created a new public repo and also created a gitignore file based on the terraform template. So the .gitignore file exists in the online repo.</li>
<li>Created a .gitignore file in my local project folder and copied the contents from the online file in the repo.</li>
<li>Performed a push and it still complains about the large .exe file in .terraform.</li>
</ol>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/0/c/0c1b5d3bb216b6e75da8e6559448685032c6708b.png" data-download-href="/uploads/short-url/1J6m8UqCRuE8pCygLFXJPZv4XSr.png?dl=1" title="vscode" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/84232822/181074590-0cd6b6be-5c52-4e71-b192-fa29f9ddefd0.png" alt="vscode" data-base62-sha1="1J6m8UqCRuE8pCygLFXJPZv4XSr" width="690" height="421" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/0/c/0c1b5d3bb216b6e75da8e6559448685032c6708b_2_690x421.png, https://aws1.discourse-cdn.com/github/optimized/3X/0/c/0c1b5d3bb216b6e75da8e6559448685032c6708b_2_1035x631.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/0/c/0c1b5d3bb216b6e75da8e6559448685032c6708b_2_1380x842.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/0/c/0c1b5d3bb216b6e75da8e6559448685032c6708b_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">vscode</span><span class="informations">1445×883 57.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by airtower-luna** [comment]
<p>That looks like that .exe file was already committed to your repository before you added the .gitignore file. A .gitignore file can’t change the repository history, you’ll have to remove the file from history yourself. The procedure is essentially the same as for <a href="https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/removing-sensitive-data-from-a-repository" class="inline-onebox">Removing sensitive data from a repository - GitHub Docs</a>.</p>
<p>For editing repository history (and related pitfalls) in general, see: <a href="https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History" class="inline-onebox">Git - Rewriting History</a></p>

**Comment by mkx85** [comment]
<p>yes that’s definitely the issue thanks! But I really wasn’t expecting the fix to be such a headache. Thanks for the links, the first one talks about using git filter-repo, but it requires python to be installed. There must be an easier way to untrack all the initial files?</p>
<p>I’ve also tried following this, but it hasn’t worked, that pesky exe is being tracked by git: <a href="https://www.codeblocq.com/2016/01/Untrack-files-already-added-to-git-repository-based-on-gitignore/" class="inline-onebox" rel="noopener nofollow ugc">Untrack files already added to git repository based on .gitignore - CodeBlocQ</a></p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="mkx85" data-post="3" data-topic="213922">
<div class="title">
<div class="quote-controls"></div>
 mkx85:</div>
<blockquote>
<p>There must be an easier way to untrack all the initial files?</p>
</blockquote>
</aside>
<p>Unfortunately not, unless you want to start over from scratch. Git history is usually meant to be immutable, modifying it is the exception, not the rule.</p>
<aside class="quote no-group" data-username="mkx85" data-post="3" data-topic="213922">
<div class="title">
<div class="quote-controls"></div>
 mkx85:</div>
<blockquote>
<p>I’ve also tried following this, but it hasn’t worked, that pesky exe is being tracked by git:</p>
</blockquote>
</aside>
<p>That guide just adds a commit that removes the files, without cleaning up history. That way the files aren’t being tracked for the future, but are still part of the history. And when you push, all history is pushed. So that guide doesn’t help you here.</p>

**Comment by thomassibin94** [comment, marked as answer]
<p>Solution is on this page. Works 100%</p><aside class="onebox githubissue" data-onebox-src="https://github.com/hashicorp/terraform-guides/issues/92">
  <header class="source">

      <a href="https://github.com/hashicorp/terraform-guides/issues/92" target="_blank" rel="noopener nofollow ugc">github.com/hashicorp/terraform-guides</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/hashicorp/terraform-guides/issues/92" target="_blank" rel="noopener nofollow ugc">Unable to ignore .terraform before pushing to GitHub</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2019-03-12" data-time="19:26:03" data-timezone="UTC">07:26PM - 12 Mar 19 UTC</span>
      </div>

        <div class="date">
          closed <span class="discourse-local-date" data-format="ll" data-date="2019-04-21" data-time="14:00:46" data-timezone="UTC">02:00PM - 21 Apr 19 UTC</span>
        </div>

      <div class="user">
        <a href="https://github.com/nguyendviet" target="_blank" rel="noopener nofollow ugc">
          <img alt="nguyendviet" src="https://user-images.githubusercontent.com/58992555/181074593-35915f58-2e50-4bf1-9e5d-a5c82f2a31f0.png" class="onebox-avatar-inline" width="20" height="20">
          nguyendviet
        </a>
      </div>
    </div>

    <div class="labels">
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">I have an error that prevents me from pushing my changes to my GitHub repo:

`<span class="show-more-container"><a href="" rel="noopener" class="show-more">…</a></span><span class="excerpt hidden">remote: error: File .terraform/plugins/darwin_amd64/terraform-provider-aws_v2.1.0_x4 is 134.07 MB; this exceeds GitHub's file size limit of 100.00 MB`

I understand the issue is the file is too large to push to github. I've tried adding the whole .terraform folder to my .gitignore file using these guides: [from GitHub](https://github.com/github/gitignore/blob/master/Terraform.gitignore) and [from Hashicorp](https://github.com/hashicorp/terraform-guides/blob/master/.gitignore). My `.terraform` is greyed out in my VSCode.

I've also tried these from several answers on SOF:
```
git rm -r --cached &lt;filePath&gt;
git reset -r -- &lt;filePath&gt;
# rm all files
git rm -r --cached .
# add all files as per new .gitignore
git add .
# now, commit for new .gitignore to apply
git commit -m ".gitignore is now working"
```

I've also tried [this](https://help.github.com/en/articles/removing-files-from-a-repositorys-history#removing-a-file-added-in-the-most-recent-unpushed-commit).

However:
```
$ git rm --cached -r .terraform
fatal: pathspec '.terraform' did not match any files
$ git rm -r --cached ./.terraform
fatal: pathspec './.terraform' did not match any files
$ ls -la
total 216
drwxr-xr-x  15 viet  staff    480 12 Mar 15:07 .
drwxr-xr-x  16 viet  staff    512  7 Mar 14:36 ..
drwxr-xr-x  15 viet  staff    480 12 Mar 15:11 .git
-rw-r--r--   1 viet  staff     66 29 Jan 15:04 .gitattributes
-rw-r--r--   1 viet  staff    933 12 Mar 15:08 .gitignore
drwxr-xr-x   3 viet  staff     96 12 Mar 15:07 .terraform
...
$ git push
Counting objects: 35, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (33/33), done.
Writing objects: 100% (35/35), 26.51 MiB | 5.04 MiB/s, done.
Total 35 (delta 17), reused 0 (delta 0)
remote: Resolving deltas: 100% (17/17), completed with 1 local object.
remote: error: GH001: Large files detected. You may want to try Git Large File Storage - https://git-lfs.github.com.
remote: error: Trace: 9cc1ae0d9a2d56bd6f7e8b4b152a10cd
remote: error: See http://git.io/iEPt8g for more information.
remote: error: File .terraform/plugins/darwin_amd64/terraform-provider-aws_v2.1.0_x4 is 134.07 MB; this exceeds GitHub's file size limit of 100.00 MB
To github.com:nditech/NDI-Demtools.git
 ! [remote rejected] master -&gt; master (pre-receive hook declined)
error: failed to push some refs to 'git@github.com:&lt;REPO-NAME&gt;'
```</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
